### ISSUE TITLE
"Object does not exist" and other errors during cp, rm and ls

### ISSUE THREAD

**Issue by gmile** [minio/mc#2412]
**TL;DR**. Click **Failing `cp` and `rm`** below to see the flickering problems related to `cp` and `rm`. Click **Failing `ls`** below to see the flickering problems related to `ls`. This is against minio server `RELEASE.2018-02-09T23-07-36Z` (installed as [minio kubernetes chart](https://github.com/kubernetes/charts/tree/master/stable/minio) v0.5.5). Also, `ls` fails to get a list of buckets once in a while.

## Expected behaviour

* during `mc rm`, an object is successfully removed
* during `mc cp`, an object is successfully copied.

### `mc rm` / `mc cp`

I'm making a test:

1. create a hundreed of 100mb files,
2. upload files to minio server,
3. remove files from minio server.

Here's the console output:

<details>
<summary>Failing `cp` and `rm`</summary>

```
gmile@192 ~/w/ehealth.charts (master) [1]> ls /tmp/minio-testing/
capitation-report-sample     capitation-report-sample-21  capitation-report-sample-35  capitation-report-sample-49  capitation-report-sample-62  capitation-report-sample-76  capitation-report-sample-9
capitation-report-sample-1   capitation-report-sample-22  capitation-report-sample-36  capitation-report-sample-5   capitation-report-sample-63  capitation-report-sample-77  capitation-report-sample-90
capitation-report-sample-10  capitation-report-sample-23  capitation-report-sample-37  capitation-report-sample-50  capitation-report-sample-64  capitation-report-sample-78  capitation-report-sample-91
capitation-report-sample-100 capitation-report-sample-24  capitation-report-sample-38  capitation-report-sample-51  capitation-report-sample-65  capitation-report-sample-79  capitation-report-sample-92
capitation-report-sample-11  capitation-report-sample-25  capitation-report-sample-39  capitation-report-sample-52  capitation-report-sample-66  capitation-report-sample-8   capitation-report-sample-93
capitation-report-sample-12  capitation-report-sample-26  capitation-report-sample-4   capitation-report-sample-53  capitation-report-sample-67  capitation-report-sample-80  capitation-report-sample-94
capitation-report-sample-13  capitation-report-sample-27  capitation-report-sample-40  capitation-report-sample-54  capitation-report-sample-68  capitation-report-sample-81  capitation-report-sample-95
capitation-report-sample-14  capitation-report-sample-28  capitation-report-sample-41  capitation-report-sample-55  capitation-report-sample-69  capitation-report-sample-82  capitation-report-sample-96
capitation-report-sample-15  capitation-report-sample-29  capitation-report-sample-42  capitation-report-sample-56  capitation-report-sample-7   capitation-report-sample-83  capitation-report-sample-97
capitation-report-sample-16  capitation-report-sample-3   capitation-report-sample-43  capitation-report-sample-57  capitation-report-sample-70  capitation-report-sample-84  capitation-report-sample-98
capitation-report-sample-17  capitation-report-sample-30  capitation-report-sample-44  capitation-report-sample-58  capitation-report-sample-71  capitation-report-sample-85  capitation-report-sample-99
capitation-report-sample-18  capitation-report-sample-31  capitation-report-sample-45  capitation-report-sample-59  capitation-report-sample-72  capitation-report-sample-86
capitation-report-sample-19  capitation-report-sample-32  capitation-report-sample-46  capitation-report-sample-6   capitation-report-sample-73  capitation-report-sample-87
capitation-report-sample-2   capitation-report-sample-33  capitation-report-sample-47  capitation-report-sample-60  capitation-report-sample-74  capitation-report-sample-88
capitation-report-sample-20  capitation-report-sample-34  capitation-report-sample-48  capitation-report-sample-61  capitation-report-sample-75  capitation-report-sample-89
gmile@192 ~/w/ehealth.charts (master) [1]>
le@192 ~/w/ehealth.charts (master)> mc cp --recursive /tmp/minio-testing/ denovo/testing-upload/
mc: <ERROR> Failed to copy `/tmp/minio-testing/capitation-report-sample-82`. Object does not exist
mc: <ERROR> Failed to copy `/tmp/minio-testing/capitation-report-sample-83`. Object does not exist
mc: <ERROR> Failed to copy `/tmp/minio-testing/capitation-report-sample-84`. Object does not exist
mc: <ERROR> Failed to copy `/tmp/minio-testing/capitation-report-sample-86`. Object does not exist
...ting/capitation-report-sample-99:  97.81 MB / 101.00 MB  ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ░░░   96.84% 8.94 MB/s 10s
gmile@192 ~/w/ehealth.charts (master) [1]>
gmile@192 ~/w/ehealth.charts (master)> for i in (seq 100); mc rm denovo/testing-upload/capitation-report-sample-$i ; end
Removing `denovo/testing-upload/capitation-report-sample-1`.
Removing `denovo/testing-upload/capitation-report-sample-2`.
Removing `denovo/testing-upload/capitation-report-sample-3`.
Removing `denovo/testing-upload/capitation-report-sample-4`.
Removing `denovo/testing-upload/capitation-report-sample-5`.
Removing `denovo/testing-upload/capitation-report-sample-6`.
Removing `denovo/testing-upload/capitation-report-sample-7`.
Removing `denovo/testing-upload/capitation-report-sample-8`.
Removing `denovo/testing-upload/capitation-report-sample-9`.
Removing `denovo/testing-upload/capitation-report-sample-10`.
Removing `denovo/testing-upload/capitation-report-sample-11`.
Removing `denovo/testing-upload/capitation-report-sample-12`.
Removing `denovo/testing-upload/capitation-report-sample-13`.
Removing `denovo/testing-upload/capitation-report-sample-14`.
Removing `denovo/testing-upload/capitation-report-sample-15`.
Removing `denovo/testing-upload/capitation-report-sample-16`.
Removing `denovo/testing-upload/capitation-report-sample-17`.
Removing `denovo/testing-upload/capitation-report-sample-18`.
Removing `denovo/testing-upload/capitation-report-sample-19`.
Removing `denovo/testing-upload/capitation-report-sample-20`.
Removing `denovo/testing-upload/capitation-report-sample-21`.
Removing `denovo/testing-upload/capitation-report-sample-22`.
Removing `denovo/testing-upload/capitation-report-sample-23`.
Removing `denovo/testing-upload/capitation-report-sample-24`.
Removing `denovo/testing-upload/capitation-report-sample-25`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-26`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-27`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-28`.
Removing `denovo/testing-upload/capitation-report-sample-29`.
Removing `denovo/testing-upload/capitation-report-sample-30`.
Removing `denovo/testing-upload/capitation-report-sample-31`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-32`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-33`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-34`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-35`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-36`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-37`.
Removing `denovo/testing-upload/capitation-report-sample-38`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-39`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-40`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-41`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-42`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-43`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-44`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-45`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-46`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-47`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-48`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-49`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-50`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-51`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-52`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-53`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-54`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-55`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-56`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-57`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-58`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-59`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-60`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-61`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-62`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-63`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-64`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-65`.
Removing `denovo/testing-upload/capitation-report-sample-66`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-67`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-68`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-69`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-70`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-71`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-72`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-73`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-74`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-75`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-76`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-77`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-78`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-79`.
Removing `denovo/testing-upload/capitation-report-sample-80`.
Removing `denovo/testing-upload/capitation-report-sample-81`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-82`. Object does not exist
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-83`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-84`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-85`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-86`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-87`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-88`. The specified bucket does not exist.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-89`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-90`.
Removing `denovo/testing-upload/capitation-report-sample-91`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-92`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-93`.
Removing `denovo/testing-upload/capitation-report-sample-94`.
mc: <ERROR> Failed to remove `denovo/testing-upload/capitation-report-sample-95`. The specified bucket does not exist.
Removing `denovo/testing-upload/capitation-report-sample-96`.
Removing `denovo/testing-upload/capitation-report-sample-97`.
Removing `denovo/testing-upload/capitation-report-sample-98`.
Removing `denovo/testing-upload/capitation-report-sample-99`.
Removing `denovo/testing-upload/capitation-report-sample-100`.
gmile@192 ~/w/ehealth.charts (master)>
```

</details>

Note: If I run `rm` one-liner over and over again, `mc rm` will successfully remove a few more files every time.

<details>
<summary>Failing `ls`</summary>

```
gmile@192 ~/w/ehealth.charts (master) [1]> mc ls denovo/testing-upload/
mc: <ERROR> Unable to stat `denovo/testing-upload/`. Bucket `testing-upload` does not exist.
gmile@192 ~/w/ehealth.charts (master) [1]> mc ls denovo/testing-upload/
mc: <ERROR> Unable to stat `denovo/testing-upload/`. Bucket `testing-upload` does not exist.
gmile@192 ~/w/ehealth.charts (master) [1]> mc ls denovo/testing-upload/
mc: <ERROR> Unable to stat `denovo/testing-upload/`. Bucket `testing-upload` does not exist.
gmile@192 ~/w/ehealth.charts (master) [1]> mc ls denovo/testing-upload/
[2018-03-14 16:36:44 EET] 1.0MiB capitation-report-sample
[2018-03-14 16:36:52 EET] 1.0MiB capitation-report-sample-72
[2018-03-14 16:36:52 EET] 1.0MiB capitation-report-sample-74
[2018-03-14 16:36:54 EET] 1.0MiB capitation-report-sample-95
```
</details>

## Steps to reproduce the behaviour

The issue is flickering, but I will try and come up with a reproducible setup.

## mc version

<pre>
<b>Version: 2018-02-09T23:07:36Z</b>
Release-tag: RELEASE.2018-02-09T23-07-36Z
Commit-id: 3987f1405aa7a0faaff019cc3b55da4ded9ebbe5
</pre>

## System information

```
macOS: 10.12.3
minio server: RELEASE.2018-02-09T22-40-05Z (single minio instance mode aka "standalone")
```

---

The problem is flickering but can be reproduced every time I run . I'll try to gather & post some logs in a bit.

**Comment by gmile** [comment]
Just realized there's the same problem `mb`. The bucket is created the second time I call `mb`, but fails on first attempt:

```
mc mb denovo/testing-upload
mc: <ERROR> Unable to make bucket `denovo/testing-upload`. The specified bucket does not exist.
gmile@192 ~/w/ehealth.charts (master) [1]> mc mb denovo/testing-upload
Bucket created successfully `denovo/testing-upload`.
```

This starts to feel like systematic failure with my setup.

**Comment by gmile** [comment]
Here's a log of uploading via `cp` – https://gist.github.com/gmile/a8677fd66d43d01f5d248d5e13a9eb2a. CMD+F for `ERROR` string on the page.

**Comment by gmile** [comment]
During a `cp`, this is what I get whenever the files is not copied:

```json
{
  "status": "error",
  "error": {
    "message": "Failed to copy `/tmp/minio-testing/capitation-report-sample-97`.",
    "cause": {
      "message": "Object does not exist",
      "error": {}
    },
    "type": "error",
    "trace": [
      {
        "line": 621,
        "file": "client-s3.go",
        "func": "cmd.(*s3Client).Put"
      },
      {
        "line": 113,
        "file": "common-methods.go",
        "func": "cmd.putTargetStream",
        "env": {
          "Tags": [
            "denovo",
            "https://minio-denovo.ehealth.world/testing-upload/capitation-report-sample-97"
          ]
        }
      },
      {
        "line": 169,
        "file": "common-methods.go",
        "func": "cmd.uploadSourceToTargetURL",
        "env": {
          "Tags": [
            "https://minio-denovo.ehealth.world/testing-upload/capitation-report-sample-97"
          ]
        }
      },
      {
        "line": 372,
        "file": "cp-main.go",
        "func": "cmd.doCopySession",
        "env": {
          "Tags": [
            "/tmp/minio-testing/capitation-report-sample-97"
          ]
        }
      }
    ],
    "sysinfo": {
      "host.arch": "amd64",
      "host.cpus": "8",
      "host.lang": "go1.9.4",
      "host.name": "IEVGENs-MacBook-Pro.local",
      "host.os": "darwin",
      "mem.heap.total": "17MB",
      "mem.heap.used": "10MB",
      "mem.total": "22MB",
      "mem.used": "10MB"
    }
  }
}
```

**Comment by harshavardhana** [comment]
can you share your minio server configuration as well here?

**Comment by gmile** [comment]
@harshavardhana my apologies, it was a weird issue with our load balancer. We initially deployed two minio instances, then deleted one deployment, but forgot to remove it from the balancer.
